core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java

```diff
@@ -153,14 +158,24 @@ public int saveInstanceMetaData(Stack stack, Iterable<CloudVmMetaDataStatus> clo
                 if (instanceMetaDataEntry.getInstanceMetadataType() == null) {
                     if (ig != null) {
                         if (InstanceGroupType.GATEWAY.equals(ig.getInstanceGroupType())) {
-                            if (!primaryIgSelected) {
-                                primaryIgSelected = true;
-                                instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.GATEWAY_PRIMARY);
-                                instanceMetaDataEntry.setServer(Boolean.TRUE);
-                                LOGGER.info("Primary gateway is not selected, let's select this instance: {}", instanceMetaDataEntry.getInstanceId());
+                            if (terminatedPrimaryGwWhichShouldBeRestored.isPresent()) {
```

Comment:
If I want a pseudocode from what happens here is (I think):
1. we first would collect all the GATEWAY or GATEWAY_PRIMARY nodes
2. in case it's a repair and we are repairing a PGW we don't change anything 
3. if we found no PWG we elect the first one

If so can we reorganise the code according to this logic? The now 5 ifs makes this quite hard to understand and log properly.